cmd/gcp-controller-manager/app/gke_approver.go

```diff
@@ -600,3 +622,93 @@ func isNotFound(err error) bool {
+	if apierrors.IsNotFound(err) {
+		// if there is no existing Node object, return success
+		return nil
+	}
+	if err != nil {
+		// returning an error triggers a retry of this CSR.
+		// if the errors are persistent, the CSR will not be approved and the node bootstrap will hang.
+		return fmt.Errorf("error getting node %s: %v", nodeName, err)
+	}
+
+	delete, err := shouldDeleteNode(opts, node)
+	if err != nil {
+		// returning an error triggers a retry of this CSR.
+		// if the errors are persistent, the CSR will not be approved and the node bootstrap will hang.
+		return fmt.Errorf("error determining if node %s should be deleted: %v", nodeName, err)
+	}
+	if !delete {
+		// if the existing Node object does not need to be removed, return success
+		return nil
+	}
+
+	err = client.CoreV1().Nodes().Delete(nodeName, &metav1.DeleteOptions{Preconditions: metav1.NewUIDPreconditions(string(node.UID))})
+	if apierrors.IsNotFound(err) {
+		// If we wanted to delete and the node is gone, this counts as success
+		return nil
+	}
+	if err != nil {
+		// returning an error triggers a retry of this CSR.
+		// if the errors are persistent, the CSR will not be approved and the node bootstrap will hang.
+		return fmt.Errorf("error deleting node %s: %v", nodeName, err)
+	}
+	csrApprovalStatus.WithLabelValues("node_deleted", metricsLabel).Inc()
+	csrApprovalLatency.WithLabelValues("node_deleted", metricsLabel).Observe(time.Since(start).Seconds())
+	return nil
+}
+
+func shouldDeleteNode(opts GCPConfig, node *v1.Node) (bool, error) {
+	// Newly created node might not have pod CIDR allocated yet.
+	if node.Spec.PodCIDR == "" {
+		glog.V(2).Infof("Node %q has empty podCIDR.", node.Name)
+		return false, nil
+	}
+	inst, err := getInstanceByName(opts, node.Name)
+	if err != nil {
+		glog.Errorf("Error retrieving instance %q: %v", node.Name, err)
+		return false, err
+	}
+	for _, networkInterface := range inst.NetworkInterfaces {
+		for _, r := range networkInterface.AliasIpRanges {
+			if node.Spec.PodCIDR == r.IpCidrRange {
+				glog.V(2).Infof("Instance %q has alias range that matches node's podCIDR.", inst.Name)
+				return false, nil
+			}
+			glog.Warningf("Instance %q has alias range %s that is different than node's podCIDR %s, will trigger node deletion.", inst.Name, r.IpCidrRange, node.Spec.PodCIDR)
+			return true, nil
+		}
+	}
+	// Instance with no alias range is route based, for which node object deletion is unnecessary.
+	glog.V(2).Infof("Instance %q has no alias range.", inst.Name)
+	return false, nil
+}
+
+func getInstanceByName(opts GCPConfig, instanceName string) (*compute.Instance, error) {
+	srv := compute.NewInstancesService(opts.Compute)
+	for _, z := range opts.Zones {
+		inst, err := srv.Get(opts.ProjectID, z, instanceName).Do()
+		if err != nil {
+			if isNotFound(err) {
+				continue
+			}
+			return nil, err
+		}
+		return inst, nil
+	}
+	return nil, fmt.Errorf("instance not found")
```

Comment:
if there is no instance in metadata matching the existing Node API object, and we return an error here and retry handling the CSR, wouldn't we just encounter the same error again?

Reply:
Don't have particular reason here --- I was simply thinking such case shouldn't happen. If we really reach here, something must be totally wrong (either kubelet got the wrong name in cert, or GCE messes up its metadata?)...But I agreed what you suggest might be a better way to handle it, will update commit.

Reply:
Updated.